### PR TITLE
Faster bitmap scans

### PR DESCRIPTION
**Pull request by richardstartin**
## Description
This uses better APIs from RoaringBitmap to speed up scans by around 25%:

At 67f5617f2c7a09b44b448bbedb610229f759d336 (the addition of the benchmark) we have:

```
Benchmark                                    (_bitmapQuantile)      (_distribution)  (_numDocs)  (_seed)  (_thresholdQuantile)  Mode  Cnt   Score   Error  Units
BenchmarkScanDocIdIterators.benchmarkSVLong                  1  UNIFORM(0,10000000)    10000000       42                     1  avgt    5  19.856 ± 0.839  ms/op
BenchmarkScanDocIdIterators.benchmarkSVLong                  1  UNIFORM(0,10000000)    10000000       42                     2  avgt    5  20.061 ± 1.085  ms/op
BenchmarkScanDocIdIterators.benchmarkSVLong                  2  UNIFORM(0,10000000)    10000000       42                     1  avgt    5  41.006 ± 1.170  ms/op
BenchmarkScanDocIdIterators.benchmarkSVLong                  2  UNIFORM(0,10000000)    10000000       42                     2  avgt    5  34.342 ± 0.271  ms/o
```

At 8a0116404b54c528131dfc6b1be6157d48317982 (the improvement) we have:

```
Benchmark                                    (_bitmapQuantile)      (_distribution)  (_numDocs)  (_seed)  (_thresholdQuantile)  Mode  Cnt   Score   Error  Units
BenchmarkScanDocIdIterators.benchmarkSVLong                  1  UNIFORM(0,10000000)    10000000       42                     1  avgt    5  15.022 ± 1.026  ms/op
BenchmarkScanDocIdIterators.benchmarkSVLong                  1  UNIFORM(0,10000000)    10000000       42                     2  avgt    5  14.576 ± 1.762  ms/op
BenchmarkScanDocIdIterators.benchmarkSVLong                  2  UNIFORM(0,10000000)    10000000       42                     1  avgt    5  35.679 ± 1.913  ms/op
BenchmarkScanDocIdIterators.benchmarkSVLong                  2  UNIFORM(0,10000000)    10000000       42                     2  avgt    5  25.985 ± 0.547  ms/op
```

At 11871a0ccc38f4a7e52d8af5147c1126c8a4b191 (better result size prediction) we have:

```
Benchmark                                    (_bitmapQuantile)      (_distribution)  (_numDocs)  (_seed)  (_thresholdQuantile)  Mode  Cnt   Score   Error  Units
BenchmarkScanDocIdIterators.benchmarkSVLong                  1  UNIFORM(0,10000000)    10000000       42                     1  avgt    5  14.498 ± 0.856  ms/op
BenchmarkScanDocIdIterators.benchmarkSVLong                  1  UNIFORM(0,10000000)    10000000       42                     2  avgt    5  14.548 ± 0.930  ms/op
BenchmarkScanDocIdIterators.benchmarkSVLong                  2  UNIFORM(0,10000000)    10000000       42                     1  avgt    5  33.848 ± 0.543  ms/op
BenchmarkScanDocIdIterators.benchmarkSVLong                  2  UNIFORM(0,10000000)    10000000       42                     2  avgt    5  25.450 ± 0.264  ms/op
```

Note that there is an aspect of space time tradeoff here, in that we allocate a one kilobyte buffer per column scan, but this is negligible compared to materialisation of a bitmap.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

**Review by kishoreg — APPROVED** [review]
LGTM other than minor comment on why 256

File: pinot-core/src/main/java/org/apache/pinot/core/operator/dociditerators/SVScanDocIdIterator.java [hunk 1]
```diff
@@ -72,16 +73,23 @@ public int advance(int targetDocId) {
...
-    IntIterator docIdIterator = docIds.getIntIterator();
-    int nextDocId;
-    while (docIdIterator.hasNext() && (nextDocId = docIdIterator.next()) < _numDocs) {
-      _numEntriesScanned++;
-      if (_valueMatcher.doesValueMatch(nextDocId)) {
-        result.add(nextDocId);
+    RoaringBitmapWriter<MutableRoaringBitmap> result = RoaringBitmapWriter.bufferWriter()
+        .expectedRange(0, _numDocs).runCompress(false).get();
+    BatchIterator docIdIterator = docIds.getBatchIterator();
+    int[] buffer = new int[256];
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/dociditerators/SVScanDocIdIterator.java]: done https://github.com/apache/pinot/pull/7530/commits/6c016cb165d3a28f31fd20e39e323a47c9530b5b

File: pinot-core/src/main/java/org/apache/pinot/core/operator/dociditerators/MVScanDocIdIterator.java [hunk 2]
```diff
@@ -73,17 +74,27 @@ public int advance(int targetDocId) {
...
+    }
+    RoaringBitmapWriter<MutableRoaringBitmap> result = RoaringBitmapWriter.bufferWriter()
+        .expectedRange(docIds.first(), docIds.last()).runCompress(false).get();
+    BatchIterator docIdIterator = docIds.getBatchIterator();
+    int[] buffer = new int[OPTIMAL_ITERATOR_BATCH_SIZE];
+    while (docIdIterator.hasNext()) {
+      int limit = docIdIterator.nextBatch(buffer);
+      for (int i = 0; i < limit; i++) {
+        int nextDocId = buffer[i];
+        if (nextDocId < _numDocs) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/dociditerators/MVScanDocIdIterator.java:88]: (Optional and not introduced in this PR) I think this if check is redundant. We need this check for the old star-tree, which is already removed and no longer supported.

**Review by Jackie-Jiang — APPROVED** [review]
LGTM. Great optimization!

File: pinot-core/src/main/java/org/apache/pinot/core/operator/dociditerators/MVScanDocIdIterator.java (hunk 2 above)
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/dociditerators/MVScanDocIdIterator.java]: good suggestion, done.

**Pull request merged**